File: src/presentation/pages/processing_page.js
import { bottomNavShell, mountBottomNav } from '../widgets/bottom_nav_shell.js';
import { processingStepTile } from '../widgets/processing_step_tile.js';

const STEPS = [
  { id: 1, label: 'Validating sensor data' },
  { id: 2, label: 'Calculating FreeAcc' },
  { id: 3, label: 'Detecting gait events' },
  { id: 4, label: 'Calculating semiogram' },
  { id: 5, label: 'Predicting FMA-LE' },
  { id: 6, label: 'Saving result' },
];

const STEP_DURATION_MS = 1200;

let _timers = [];

const PROFILE_IMG = 'https://lh3.googleusercontent.com/aida-public/AB6AXuCl8V4EMSiz6Thf6H7WehwuLc1j3OGBJipEKTCeBPMwc1ytiZusOmgAxUkgTvcy0wm01qq7FYIR8Pc3rNWhGThjsXuMMVsMfaUAtRSqa7e9EQpIfdXuFn62o3VgyMlaXe_1CksuuY1uTnnRW6FL63zduph3YUXjdVI2HshdG5K0XDsH9hVTl-5voAUDSAxc6BJHSS7u-xIee8BDMsqSGOFK37VpCOM7Uw7T9zRoT9BG5MuE5Oh2anK9qz7p7wYkElL_gpwrm8SIiaA';

const WAVE_BARS = [20, 35, 50, 40, 25, 45, 30].map((h, i) =>
  `<div class="wave-bar" style="height:${h}px;animation-delay:${i * 0.17}s;"></div>`
).join('');

export function render() {
  return `
    <div class="min-h-screen flex flex-col font-body-md text-on-surface" style="background-color:#f4fbf8;">
      <!-- Top App Bar -->
      <header class="bg-surface/80 backdrop-blur-md shadow-sm fixed top-0 w-full z-50 flex justify-between items-center px-margin-mobile h-16 transition-all duration-200">
        <div class="flex items-center gap-md">
          <div class="w-10 h-10 rounded-full bg-surface-container-highest overflow-hidden border border-outline-variant/30">
            <img src="${PROFILE_IMG}" alt="User" class="w-full h-full object-cover">
          </div>
          <h1 class="font-headline-sm text-headline-sm text-primary">Gait Analysis</h1>
        </div>
        <div class="flex items-center">
          <span class="material-symbols-outlined text-primary text-2xl">sensors</span>
        </div>
      </header>

      <main class="flex-grow pt-24 pb-32 px-margin-mobile max-w-md mx-auto w-full">
        <!-- Hero -->
        <section class="flex flex-col items-center text-center mb-xl">
          <div class="w-48 h-48 bg-surface-container-lowest rounded-full shadow-[0_4px_20px_0_rgba(29,53,87,0.06)] flex items-center justify-center mb-lg relative overflow-hidden">
            <div class="wave-container">${WAVE_BARS}</div>
            <div class="absolute inset-0 flex items-center justify-center pointer-events-none opacity-10">
              <span class="material-symbols-outlined text-8xl text-primary">insights</span>
            </div>
          </div>
          <h2 class="font-headline-md text-headline-md text-on-surface mb-sm">Processing Your Gait Data</h2>
          <p class="font-body-md text-on-surface-variant max-w-[280px]">Please wait while we analyze your walking pattern.</p>
        </section>

        <!-- Processing Stepper -->
        <div class="bg-surface-container-lowest rounded-lg p-lg shadow-[0_4px_20px_0_rgba(29,53,87,0.06)] border border-outline-variant/10">
          <div class="flex flex-col gap-lg relative">
            <div class="absolute left-3.5 top-2 bottom-2 w-0.5 bg-outline-variant/30"></div>
            ${STEPS.map(s => processingStepTile({ id: s.id, label: s.label, status: 'pending' })).join('')}
          </div>
        </div>

        <!-- Tip -->
        <div class="mt-xl text-center">
          <div class="inline-flex items-center gap-xs px-md py-sm bg-secondary-container/20 rounded-full">
            <span class="material-symbols-outlined text-secondary text-sm">lightbulb</span>
            <span class="font-label-sm text-label-sm text-on-secondary-container">Tip: Keep your phone steady for better accuracy</span>
          </div>
        </div>
      </main>

      ${bottomNavShell('record')}
    </div>
  `;
}

export function mount(navigate) {
  _timers = [];
  mountBottomNav(navigate);

  STEPS.forEach((step, i) => {
    _timers.push(setTimeout(() => setStepStatus(step, 'active'),    i * STEP_DURATION_MS));
    _timers.push(setTimeout(() => setStepStatus(step, 'completed'), i * STEP_DURATION_MS + STEP_DURATION_MS * 0.6));
  });

  _timers.push(setTimeout(() => navigate('/result'), STEPS.length * STEP_DURATION_MS + 600));
}

export function unmount() {
  _timers.forEach(clearTimeout);
  _timers = [];
}

function setStepStatus(step, status) {
  const el = document.getElementById(`step-${step.id}`);
  if (!el) return;
  el.outerHTML = processingStepTile({ id: step.id, label: step.label, status });
}

File: lib/presentation/pages/processing_page.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';
import 'package:provider/provider.dart';
import '../../core/constants/app_colors.dart';
import '../../core/constants/app_routes.dart';
import '../../core/constants/app_text_styles.dart';
import '../../core/theme/app_theme.dart';
import '../../core/utils/dummy_data.dart';
import '../../data/services/api_service.dart';
import '../providers/session_provider.dart';
import '../providers/recording_provider.dart';
import '../providers/profile_provider.dart';
import '../widgets/bottom_nav_shell.dart';

class ProcessingPage extends StatefulWidget {
  const ProcessingPage({super.key});

  @override
  State<ProcessingPage> createState() => _ProcessingPageState();
}

class _ProcessingPageState extends State<ProcessingPage>
    with SingleTickerProviderStateMixin {
  int _currentStep = 0;
  late AnimationController _waveController;

  static const List<String> _steps = [
    'Validating sensor data',
    'Calculating FreeAcc',
    'Detecting gait events',
    'Calculating semiogram',
    'Predicting FMA-LE',
    'Saving result',
  ];

  @override
  void initState() {
    super.initState();
    _waveController = AnimationController(
      vsync: this,
      duration: const Duration(seconds: 2),
    )..repeat();
    _runSteps();
  }

  void _runSteps() {
    // Cosmetic step animation while the real backend call runs.
    for (int i = 0; i < _steps.length; i++) {
      Future.delayed(Duration(milliseconds: 500 + i * 600), () {
        if (mounted && _currentStep < i + 1) {
          setState(() => _currentStep = i + 1);
        }
      });
    }
    _analyze();
  }

  Future<void> _analyze() async {
    final recording = context.read<RecordingProvider>();
    final profile = context.read<ProfileProvider>().profile;
    final session = context.read<SessionProvider>();
    final messenger = ScaffoldMessenger.of(context);

    var result = dummyResult;
    String? error;
    try {
      result = await ApiService().analyzeGaitSession(
        timeseries: recording.recordedData,
        profile: profile,
        samplingRateHz: recording.samplingRateHz,
      );
    } on ApiException catch (e) {
      error = e.message;
    } catch (e) {
      error = e.toString();
    }
    if (!mounted) return;

    // Let the step animation breathe a little before navigating.
    if (_currentStep < _steps.length) {
      setState(() => _currentStep = _steps.length);
    }
    await Future.delayed(const Duration(milliseconds: 400));
    if (!mounted) return;

    session.setResult(result);
    if (error != null) {
      messenger.showSnackBar(SnackBar(
        content: Text(
            'Backend unavailable — showing sample result. $error'),
        duration: const Duration(seconds: 4),
      ));
    }
    context.go(AppRoutes.result);
  }

  @override
  void dispose() {
    _waveController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppColors.background,
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(AppTheme.marginMobile),
          child: Column(
            children: [
              const SizedBox(height: AppTheme.xl),
              // Hero wave animation
              _WaveHero(controller: _waveController),
              const SizedBox(height: AppTheme.xl),
              Text('Processing Your Gait Data',
                  style: AppText.headlineMd,
                  textAlign: TextAlign.center),
              const SizedBox(height: AppTheme.sm),
              Text(
                'Our AI is analyzing your movement patterns.\nThis may take a moment.',
                style: AppText.bodyMd
                    .copyWith(color: AppColors.onSurfaceVariant),
                textAlign: TextAlign.center,
              ),
              const SizedBox(height: AppTheme.xl),
              // Steps card
              Container(
                decoration: BoxDecoration(
                  color: AppColors.surfaceContainerLowest,
                  borderRadius: AppTheme.radiusDefault,
                  border: Border.all(
                      color: AppColors.outlineVariant.withAlpha(26)),
                  boxShadow: AppTheme.cardShadow,
                ),
                padding: const EdgeInsets.all(AppTheme.lg),
                child: Column(
                  children: List.generate(
                    _steps.length,
                    (i) => _StepTile(
                      label: _steps[i],
                      state: i < _currentStep
                          ? _StepState.done
                          : i == _currentStep
                              ? _StepState.active
                              : _StepState.pending,
                      isLast: i == _steps.length - 1,
                    ),
                  ),
                ),
              ),
              const SizedBox(height: AppTheme.lg),
              // Tip chip
              Container(
                padding: const EdgeInsets.symmetric(
                    horizontal: AppTheme.md, vertical: AppTheme.sm),
                decoration: BoxDecoration(
                  color: AppColors.primaryContainer.withAlpha(26),
                  borderRadius: AppTheme.radiusFull,
                ),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    const Icon(Icons.lightbulb_outline,
                        size: 16, color: AppColors.primary),
                    const SizedBox(width: 6),
                    Text('Keep your phone steady during processing',
                        style: AppText.labelSm
                            .copyWith(color: AppColors.primary)),
                  ],
                ),
              ),
              const SizedBox(height: 80),
            ],
          ),
        ),
      ),
      bottomNavigationBar: const BottomNavBar(activePage: 'record'),
    );
  }
}

class _WaveHero extends StatelessWidget {
  final AnimationController controller;
  const _WaveHero({required this.controller});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 192,
      height: 192,
      child: Stack(
        alignment: Alignment.center,
        children: [
          // Pulsing rings
          AnimatedBuilder(
            animation: controller,
            builder: (_, __) => Stack(
              alignment: Alignment.center,
              children: [
                _PulseRing(
                    scale: 1.0 + controller.value * 0.3,
                    opacity: 1.0 - controller.value),
                _PulseRing(
                    scale: 1.0 + ((controller.value + 0.5) % 1.0) * 0.3,
                    opacity: 1.0 - (controller.value + 0.5) % 1.0),
              ],
            ),
          ),
          // Core circle
          Container(
            width: 120,
            height: 120,
            decoration: const BoxDecoration(
              gradient: LinearGradient(
                begin: Alignment.topLeft,
                end: Alignment.bottomRight,
                colors: [AppColors.primary, AppColors.primaryContainer],
              ),
              shape: BoxShape.circle,
            ),
            child: const Icon(
              Icons.analytics_outlined,
              size: 48,
              color: Colors.white,
            ),
          ),
        ],
      ),
    );
  }
}

class _PulseRing extends StatelessWidget {
  final double scale;
  final double opacity;
  const _PulseRing({required this.scale, required this.opacity});

  @override
  Widget build(BuildContext context) {
    return Transform.scale(
      scale: scale,
      child: Container(
        width: 140,
        height: 140,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          border: Border.all(
            color: AppColors.primaryContainer.withAlpha((opacity * 100).toInt()),
            width: 2,
          ),
        ),
      ),
    );
  }
}

enum _StepState { done, active, pending }

class _StepTile extends StatelessWidget {
  final String label;
  final _StepState state;
  final bool isLast;

  const _StepTile({
    required this.label,
    required this.state,
    required this.isLast,
  });

  @override
  Widget build(BuildContext context) {
    final Color iconBg = switch (state) {
      _StepState.done => AppColors.primaryContainer,
      _StepState.active => AppColors.primary,
      _StepState.pending => AppColors.surfaceContainerHighest,
    };
    final Color textColor = switch (state) {
      _StepState.done => AppColors.onSurface,
      _StepState.active => AppColors.primary,
      _StepState.pending => AppColors.onSurfaceVariant,
    };

    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Column(
          children: [
            AnimatedContainer(
              duration: const Duration(milliseconds: 300),
              width: 32,
              height: 32,
              decoration: BoxDecoration(
                color: iconBg,
                shape: BoxShape.circle,
              ),
              child: state == _StepState.done
                  ? const Icon(Icons.check,
                      size: 16, color: AppColors.onPrimaryContainer)
                  : state == _StepState.active
                      ? const SizedBox(
                          width: 16,
                          height: 16,
                          child: Padding(
                            padding: EdgeInsets.all(8),
                            child: CircularProgressIndicator(
                              strokeWidth: 2,
                              valueColor: AlwaysStoppedAnimation<Color>(
                                  AppColors.onPrimary),
                            ),
                          ),
                        )
                      : null,
            ),
            if (!isLast)
              Container(
                width: 2,
                height: 24,
                color: state == _StepState.done
                    ? AppColors.primaryContainer
                    : AppColors.outlineVariant.withAlpha(80),
              ),
          ],
        ),
        const SizedBox(width: AppTheme.sm),
        Expanded(
          child: Padding(
            padding: EdgeInsets.only(
                top: 6, bottom: isLast ? 0 : AppTheme.sm),
            child: Text(label,
                style: AppText.labelMd.copyWith(color: textColor)),
          ),
        ),
      ],
    );
  }
}
